src/multiformats/address/codec.cljc

```diff
@@ -0,0 +1,165 @@
+           (throw (ex-info (str "Value doesn't conform to IPv" version)
+                           {:value addr :ip-version version})))
+         (let [addr-num (.toInteger ip)]
+           (dotimes [piece-idx num-pieces]
+             ;; goog.math.Integer int32 pieces are little endian
+             ;; so need to iterate backwards for network order
+             (let [rev-piece-idx (dec (- num-pieces piece-idx))
+                   piece (.getBitsUnsigned addr-num rev-piece-idx)
+                   piece-offset (* 4 piece-idx)]
+               (write-uint-bytes! piece dst piece-offset 4))))
+         dst)))
+  (bytes->str [this data]
+    (when (not= (alength ^bytes data) (* num-components bytes-per-component))
+      (throw (ex-info (str "Incorrect bytes for IPv" version)
+                      {:value (b/byte-seq data)
+                       :ip-version version})))
+    #?(:clj
+       (-> ^bytes data InetAddress/getByAddress .getHostAddress)
+       :cljs
+       ;; goog.net.IpAddress doesn't support an easy way
+       ;; to build IP from bytes, so manually construct string
+       (str/join separator
+                 (for [idx (range num-components)
+                       :let [offset (* idx bytes-per-component)]]
+                   (-> data
+                       (big-endian-bytes->uint offset bytes-per-component)
+                       (int->str radix))))))
+  (fixed-byte-length [this]
+    (* num-components bytes-per-component)))
+
+
+(def ip4-codec
+  (->IPAddressCodec 4 10 "." 4 1))
+
+(def ip6-codec
+  (->IPAddressCodec 6 16 ":" 8 2))
+
+(defn- min-bytes-for-uint [n]
+  (loop [bytes-shift 1]
+    (let [shifted (unsigned-bit-shift-right n (* 8 bytes-shift))]
+      (if (zero? shifted)
+        bytes-shift
+        (recur (inc bytes-shift))))))
+
+(defrecord UnsignedNumTranscoder [radix num-bytes]
+  Codec
+  (str->bytes [this value]
+    (let [dst (b/byte-array num-bytes)
+          n (str->int value)]
+      (when (> (min-bytes-for-uint n) num-bytes)
+        (throw (ex-info (str "Value too big for " num-bytes " bytes")
+                        {:value value :num-bytes num-bytes})))
+      (write-uint-bytes! n dst 0 num-bytes)
+      dst))
+  (bytes->str [this data]
+    (when (not= (alength ^bytes data) num-bytes)
+      (throw (ex-info "Incorrect number of bytes"
+                      {:num-bytes-given (alength ^bytes data)
+                       :expected-bytes num-bytes})))
+    (-> data big-endian-bytes->uint (int->str radix)))
+  (fixed-byte-length [this]
+    num-bytes))
+
+(def ushort-codec
+  (->UnsignedNumTranscoder 10 2))
+
+(def no-value-codec
+  (reify Codec
+    (str->bytes [this value]
+      (when-not (nil? value)
+        (throw (ex-info "Expected nil value for no value protocol"
+                        {:value value})))
+      (b/byte-array 0))
+    (bytes->str [this data]
+      (when (pos? (alength ^bytes data))
+        (throw (ex-info "Expected 0 length array for no value protocol"
```

Comment:
Ah, some of the zero-byte array stuff above makes more sense with these checks.

Reply:
Ok, I actually left this in place defensively, but changed the call-site code to not bother creating 0 byte arrays